Add documentation for reviewers and a checklist for PRs (#315)
physiopy/phys2bids: docs/contributorfile.rst

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
We use `readthedocs <https://readthedocs.org/>`_ to create our documentation. Every contribution is welcome and it follows the same steps as a code contribution, explained below.

.. _code:

Contributing code through GitHub
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
This section covers 90% of the contributions a project like ``physiopy`` receives - code, documentation and tests.
The best way to make this kind of contribution, in a nutshell, is to:
1. Open an issue with the intended modifications.
2. Label it, discuss it, (self-)assign it.
3. Open a Pull Request (PR) to resolve the issue and label it.
4. Wait for a review, discuss it or comply, repeat until ready.
Issues and PR chats are great to maintain track of the conversation on the contribution. They are based upon GitHub-flavoured `Markdown <https://daringfireball.net/projects/markdown>`_. GitHub has a helpful page on `getting started with writing and formatting Markdown on GitHub <https://help.github.com/articles/getting-started-with-writing-and-formatting-on-github>`_.

.. _reviews:

Contributing with Pull Requests Reviews
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> One big challenge in this kind of software development is merging code accurately but without waiting too much time.

Reviewer: ```suggestion
A big challenge of software development is merging code accurately but without having to wait too much time.
```

Labels
------
The current list of labels are `here <https://github.com/physiopy/phys2bids/labels>`_. They can be used for **Issues**, **PRs**, or both.
We use `auto <https://github.com/intuit/auto>`_ to automate our semantic versioning and Pypi upload, so **it's extremely important to use the right PR labels**!

.. _issueprlabels:

Issue & PR labels
~~~~~~~~~~~~~~~~~~~

> -  Documentation: Improvements or additions to documentation. This category includes (but is not limited to) docs pages, docstrings, and code comments.

Reviewer 2: This is a minor thing, but it would be awesome if the labels themselves could be replaced with badges.

Reviewer 2: Here's an example (built [here](https://shields.io/category/issue-tracking)): https://img.shields.io/github/labels/physiopy/phys2bids/Documentation?style=plastic

Reviewer: > This is a minor thing, but it would be awesome if the labels themselves could be replaced with badges.

You mean in the bullet list?

Reviewer 2: Yeah, just in the documentation.

Reviewer 2: And the badges could probably link to the list of issues with each label, if we want.

Reviewer: That would be very convenient actually! I like that.

Author: Yeah I wondered how did you do it in `tedana`, cause I definitely wanted to do it here!
I wouldn't stall this PR for this reason, but we can definitely open an issue about it!

Author: (Also I would use the _flat_ style rather than the plastic one)

In general, if you're tempted to assign more than one label that would trigger a release, you might want to split your PR instead.
When opening a pull request, assign it at least one label.

We encourage you to open a PR as soon as possible - even before you finish working on them. This is useful especially to you - so that you can receive comments and suggestions early on, rather than having to process a lot of comments in the final review step!
However, if it’s an incomplete PR, please open a **Draft PR**. That helps us process PRs by knowing which one to have a look first - and how picky to be when doing so.

Reviewing PRs is a time consuming task, that can be stressful for both the reviewer and the author. Avoiding wasting time and the need of little fixes - such as fixing grammar mistakes and typos, styling code, or adopting conventions - is a good start for a successful (and quick) review. Before graduating a Draft PR to a PR ready for review, please check that:

- You did all you wanted to include in your PR. If at a later stage you realise something is missing and it's not a minor thing, you will need to open a new PR.

> - If your contribution contains code or tests, you run and passed all the tests locally with `pytest`.

Reviewer 2: ```suggestion
- If your contribution contains code or tests, you ran and passed all of the tests locally with `pytest`.
```

We invite more expert contributors to avoid those issues, leave them to beginners and possibly help them out in the resolution of the issue. However, if the issue is left unassigned or unattended for long, and it’s considered important or urgent, anyone can tackle it.

.. _workflow:

Contribution workflow
---------------------
There are many descriptions of a good contribution workflow out there. For instance, we suggest to have a look at `tedana's workflow <https://github.com/ME-ICA/tedana/blob/master/CONTRIBUTING.md#making-a-change>`_.
At ``physiopy``, we follow a very similar workflow. The only three differences are:

- If you see an open issue that you would like to work on, check if it is assigned. If it is, ask the assignee if they need help or want to be substituted before starting to work on it.
- We ask you to test the code locally before merging it, and then, if possible, write some automatic tests for the code to be run in our Continuous Integration! Check the testing section below to know more.
- We suggest opening a draft PR as soon as you can - so it’s easier for us to help you!

.. _pr:

Pull Requests
-------------

> To improve understanding pull requests "at a glance", we use the labels listed above. Multiple labels can be assigned, but it's a good idea to keep different types of contributions separated, unless they are minimal - for instance, you might want to open a PR for code, one for documentation and one for testing.

Reviewer 2: This confuses me a little. It implies to me that one set of changes (e.g., adding a new metric to `phys2denoise`), would involve three PRs: one to add the metric, one to add documentation about the metric, and one to add tests for the metric. The last one clearly falls under the first PR based on the rest of this document, but I don't think it's clear from context if contributors should separate the PR with the new metric+testing from another PR with documentation about the metric. Personally, I think all of it should be done in one PR in that situation.

Author: The reasoning is two(three?)fold.
First, if we want to use the automatic `all-contributors` plugin for `auto`, if reviewers use the "suggestion" tool they are interpreted as co-authors of the PR. For instance, this PR is labelled "internal" and "documentation", so all reviewers I accepted suggestion from will get the contribution badge for `documentation` and `infrastructure`. Normally it doesn't matter so much because we tend to review the types of PR that match our previous contribution, but I already have to clean up a bit the contribution table.
Second, since the main focus of a `code` PR is code changes (and not tests or documentations, that have to be done but are _secondary_ in that context), even if the PR contains changes to the tests and documentation I would only label it with one of the two code-related label (majormod and minormod), since the code change is the real scope of the PR. This is also related to the fact that all-contributors has a binary nature (either you contributed in a field or you didn't). I find slightly unfair that two contributors have the same badge (e.g. `test`), but one made almost all the tests and the other two small changes. This is partially solved by the number and size of the PRs labelled in a certain way, so having clear, distinct PRs helps adding a metric of contribution.
Bonus reason: we could _not_ use the all-contributors plugin for `auto`, but most of the time it helps adding contributions impartially and it really helps keeping track of most of the contributions.

Reviewer: Gotcha! Still, I think tests could go with code changes into the same PR. If we make two different PRs we will increase our version with each of them, which does not make sense to me if the PRs are 100% related like code + tests.

Reviewer 3: I think it might make sense to include tests alongside the code, in the same PR, as well. 

And if a contributor writes the code, but is not sure about how to write the tests, that is when they can make a draft PR and ask for help from other people to include tests in their PR?

Author: I'm going to update after the discussion in the call!

Reviewing PRs is a time consuming task, that can be stressful for both the reviewer and the author. Avoiding wasting time and the need of little fixes - such as fixing grammar mistakes and typos, styling code, or adopting conventions - is a good start for a successful (and quick) review. Before graduating a Draft PR to a PR ready for review, please check that:

- You did all you wanted to include in your PR. If at a later stage you realise something is missing and it's not a minor thing, you will need to open a new PR.
- If your contribution contains code or tests, you ran and passed all of the tests locally with `pytest`.
- If you're writing documentation, you built it locally with `sphinx` and the format is what you intended.
- Your code is harmonious with the rest of the code - no repetitions of any sort!
- Your code respects the `adopted Style <#styling>`_, especially:
    - Your code is lintered adequately and respects the `PEP8 <https://www.python.org/dev/peps/pep-0008/>`_ convention.
    - Your docstrings follow the `numpydoc <https://numpydoc.readthedocs.io/en/latest/format.html>`_ convention.

>     - There is no grammar mistake or typo and the text is fluid.

Reviewer: I think this one should be inlcuded.

3. Have at least a release-related label (or a `Skip release` label).
4. Have a short title that clearly explains in one sentence the aim of the PR.
5. Contain at least a unit test for your contribution, if the PR contains code (it would be better if it contains an integration or function test and all the breaking tests necessary). If you’re not confident about writing tests, it is possible to refer to an issue that asks for the test to be written, or another (Draft) PR that contains the tests required.

As we’re trying to maintain at least a 90% code coverage, you’re strongly encouraged to write all the necessary tests not to drop below the threshold. If our coverage becomes too low, you might be asked to add more tests and/or your PR might be rejected. See the `Automatic Testing <#testing>`_ section.

Don't merge your own pull request! That's a task for the main reviewer of your PR or the project manager. Remember that the project manager doesn't have to be a reviewer of your PR!

.. _reviewing:

Reviewing PRs
-------------

> Reviewing PRs is an extremely important task in collaborative development. In fact, it is probably the task that requires the most time in the development, and it can be stressful for both the reviewer and the author. Remember that, as a PR Reviewer, you are granting that the codes works and integrates well with the rest of the repository, hence **you are responsible for the quality of the repository and its next version release**. If it doesn't integrate well, later PR reviewers might have to ask for broader changes than expected.

Reviewer: I think this one should be included too.

There are many best practices to review code online, for instance `this one <https://medium.com/an-idea/the-code-review-guide-9e793edcd683>`_, but here are some good rules of thumbs that we need to follow while reviewing PRs:

- Be respectful to the PR authors and be clear in what you are asking/suggesting - remember that, like you, they are contributing their spare time and doing their best job!
- If there is a Draft PR, you can comment on its development in the message board or making "Comment" reviews. Don't ask for changes, and especially, **don't approve the PR**
- If the PR graduated from Draft to full PR, check that it follows the sections `Pull requests <#pr>`_ and `Style Guide <#styling>`_ of these guidelines. If not, invite the author to do so before starting a review.
- **Don't limit your review to the parts that are changed**. Look at the entire file, see if the changes fit well in it, and see if the changes are properly addressed everywhere in the code - in the documentation, in the tests, and in other functions. Sometimes the differences reported don't show the full impact of the PR in the repository!

> - Unless it's for typo fixes or similar, invite the author of the PR to make changes before actually doing them yourself. Request changes via comments or in the message board or by checking out the PR locally, making changes and then submitting a PR to the author's branch.

Reviewer 2: To clarify, does this mean that reviewers shouldn't use the code suggestion tool (like you probably see throughout my review) or that they shouldn't directly commit to the contributor's branch? I personally like the code suggestion tool (obviously), so I'd like to allow reviewers to use it as needed.

Author: That's a practice I liked a lot and that I tried to pick up from @rmarkello: he never used the suggestion tools, opting for explaining the changes to the code he meant to apply and waiting for the PR author to do them instead - with the exception of rewording or typos or minimal things.
If you don't agree with this practice, we can reword it so that it's only against direct commits. I know that the suggestion tool make things waaaaaaaaaay faster, but for learning purposes is not the best.

Reviewer: Maybe we could make an exception with newbies and make it educational for them. Or if one of us would like an extended description of the suggestion, we can just ask for it!

Reviewer: To me this PR is ready to be merged once we figure this part out.

Reviewer 3: I don't _think_ I have used the code suggestion tool before. I normally just make comments, or put in a PR to the author's branch, with my suggested changes. I guess it would be good to be clear about what we prefer, if anything.

Author: I'm going to rewrite this point in the next commit!

- If you are the main reviewer, and the last reviewer required to approve the PR, merge the PR!

Before approving and/or merging PRs, be sure that:

- All the tests in CircleCI/Azure pass without errors.
- Prefereably, codecov checks pass as well. If they don't, discuss what to do.
- The title describes the content of the PR clearly enough to be meaningful on its own - remember that it will appear in the version changelog!
- The PR has the appropriate labels to trigger the appropriate version release and update the contributors table.

.. _mainreviewer:

Main reviewer
~~~~~~~~~~~~~
At ``physiopy`` we use the "Assignees" section of a PR to mark the **main reviewer** for that PR. The main reviewer is the main responsible **for the quality of the repository and its next version release**, as well as **for the behaviour of the other reviewers**. 
The main reviewer:

- Takes care of the reviewing process of the PR, in particular:
    - Invites the reviewers to finish their review in a relatively short time.

>     - Checks that this document was respected in all its part, especially the part about `Pull Requests <#pr>`_.

Reviewer 2: ```suggestion
    - Checks that all elements of this document were respected, especially the part about `Pull Requests <#pr>`_.
```

Reviewer: This one should be included.

    - Invites other Reviewers to respect this document, especially the part about `reviews <#reviewing>`_, helps them in doing so, and checks that they do.
    - If a Reviewer keeps not respecting this document, flags them to the project manager.
- Decides what to do in case of a coverage decrease (in *codecov/patch*).
- In case of missing tests or updates to user documentation:
    - Asks for more documentation or tests before approving the PR, *or*

>     - Checks that the adequate issues have been opened to address the lack of documentation or tests (1 issue per item).

Reviewer 2: ```suggestion
    - Checks that the appropriate issues have been opened to address the lack of documentation or tests (1 issue per item).
```

> - Double-checks that the title is clear and the labels are correct to trigger an adequate ``auto`` release - feel free to change them.

Reviewer 2: ```suggestion
- Double-checks that the title is clear and the labels are correct to trigger an appropriate ``auto`` release - feel free to change them.
```

- **Is the one that is going to merge the PR.**

> - After the PR got merged and a new release was triggered, checks that:

Reviewer 2: ```suggestion
- After the PR has been merged and a new release has been triggered, checks that:
```

    - The documentation was updated correctly (if changed).
    - The Pypi version of the repository coincides with the new release (if changed).
    - New contributors or forms of contributions were correctly added in the README (if changed).

.. _styling:

Style Guide
-----------
Docstrings should follow `numpydoc <https://numpydoc.readthedocs.io/en/latest/format.html>`_ convention. We encourage extensive documentation.
The python code itself should follow `PEP8 <https://www.python.org/dev/peps/pep-0008/>`_ convention whenever possible: there are continuous integration tests checking that!

> You can use linters to help you write your code following style conventions. Linters are add-ons that you can run on the written script file. We suggest the use of **flake8** for Python 3. Many editors (Atom, VScode, Sublimetext, ...) support addons for online lintering, which means you’ll see warnings and errors while you write the code - check out if your does!

Reviewer: Should we wait for the discussion going on at #324 before merging?

Author: Even if we adopt an opinionated code formatter, I wouldn't force the contributors to adopt it during their coding or documentation writing. In fact, one of the reasons to adopt an opinionated code formatter is to _not have to respect any opinionated code format during coding_. Most of the time it's sufficient that the code respects *flake8* conventions, so that the changes introduced by opinionated code formatters do not introduce an incredible amount of differences compared to the submitted code - although it's true that it becomes less essential. Since we're going to have `black` run during PR merge (I guess?), I would keep this section as it is.

.. _workflow:

Contribution workflow
---------------------
There are many descriptions of a good contribution workflow out there. For instance, we suggest to have a look at `tedana's workflow <https://github.com/ME-ICA/tedana/blob/master/CONTRIBUTING.md#making-a-change>`_.
At ``physiopy``, we follow a very similar workflow. The only three differences are:

- If you see an open issue that you would like to work on, check if it is assigned. If it is, ask the assignee if they need help or want to be substituted before starting to work on it.
- We ask you to test the code locally before merging it, and then, if possible, write some automatic tests for the code to be run in our Continuous Integration! Check the testing section below to know more.
- We suggest opening a draft PR as soon as you can - so it’s easier for us to help you!

.. _pr:

Pull Requests
-------------
To improve understanding pull requests "at a glance" and use the power of ``auto``, we use the labels listed above. Multiple labels can be assigned to a PR - in fact, all those that you think are relevant.

> We strongly advise to keep the changes you're introducing with your PR limited to your original goal. Little corrections here and there in the code that you're already modifying are a great help, but they risk complicating the nature of the PR and the reviewing process. It is much better to open another PR with the objective of doing such corrections!

Reviewer: I don't agree with this. Why would we want to merge some code that we know needs corrections? We should not merge code that contains errors.

Another thing is that from that PR we think of another PR with a different goal. Then it makes sense to work on a new PR after merging.

Author: Let's see if the new rewording helps clarifying this point (commit to follow).

In general, if you're tempted to assign more than one label that would trigger a release, you might want to split your PR instead.

> When opening a pull request, assign it at least one label.

Reviewer: This is also a bit confusing. So, it's better to split the PR when assigning more than one label. But then you say the PR should have at least one label. It sounds like we're encouraging people to give more than one label with the "at least".

Author: Same, let's see if it's better now.
